lib/msal-common/src/authority/RegionDiscovery.ts

```diff
@@ -0,0 +1,91 @@
+    // Options for the IMDS endpoint request
+    protected static IMDS_OPTIONS = {headers: {"Metadata": "true"}};
+    // The version of the IMDS endpoint to use
+    protected static VERSION = "2020-06-01";
+    // Region variable name
+    protected static REGION_ENV_NAME = "REGION_NAME";
+
+    constructor(networkInterface: INetworkModule) {
+        this.networkInterface = networkInterface;
+    }
+
+    /**
+     * Detect the region from the application's environment.
+     * 
+     * @returns Promise<string | null>
+     */
+    public async detectRegion(): Promise<string | null> {
+        // Detect region from the process environment variable
+        let autodetectedRegionName = process.env[RegionDiscovery.REGION_ENV_NAME];
+
+        // Call the local IMDS endpoint for applications running in azure vms
+        if (!autodetectedRegionName) {
+            try {
+                const response = await this.networkInterface.sendGetRequestAsync<string>(this.buildIMDSEndpoint(RegionDiscovery.IMDS_ENDPOINT, RegionDiscovery.VERSION), RegionDiscovery.IMDS_OPTIONS);
+                if (response.status === 200) {
+                    autodetectedRegionName = response.body;
+                } 
+                
+                if (response.status === 400) {
+                    const latestIMDSVersion = await this.getCurrentVersion();
+                    if (!latestIMDSVersion) {
+                        return null;
+                    }
+
+                    const response = await this.networkInterface.sendGetRequestAsync<string>(this.buildIMDSEndpoint(RegionDiscovery.IMDS_ENDPOINT, latestIMDSVersion), RegionDiscovery.IMDS_OPTIONS);
+                    if (response.status === 200) {
+                        autodetectedRegionName = response.body;
+                    }
+                } 
+            } catch(e) {
+                return null;
+            } 
+        }
+
+        return autodetectedRegionName || null;
+    }
+
+    /**
+     * Construct the IMDS endpoint url
+     * 
+     * @param imdsEndpoint 
+     * @param version 
+     * @returns string 
+     */
+    private buildIMDSEndpoint(imdsEndpoint: string, version: string): string {
+        return `${imdsEndpoint}?api-version=${version}&format=text`;
+    }
+
+    /**
+     * Get the most recent version of the IMDS endpoint available
+     *  
+     * @returns Promise<string | null>
+     */
+    private async getCurrentVersion(): Promise<string | null> {
+        try {
+            const response = await this.networkInterface.sendGetRequestAsync<IMDSBadResponse>(`${RegionDiscovery.IMDS_ENDPOINT}?format=json`, RegionDiscovery.IMDS_OPTIONS);
+
+            // When IMDS endpoint is called without the api version query param, bad request response comes back with latest version.
+            if (response.status === 400) {
+                return response.body["newest-versions"][0];
```

Comment:
Please add validation that response.body is truthy and contains "newest-versions" which has at least 1 member